x448/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+        PublicKey(point)
+    }
+}
+
+/// A PublicKey is a point on Curve448.
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub struct PublicKey(MontgomeryPoint);
+
+/// A Secret is a Scalar on Curve448.
+#[derive(Clone, Zeroize)]
+#[zeroize(drop)]
+pub struct Secret([u8; 56]);
+
+/// A SharedSecret is a point on Curve448.
+/// This point is the result of a Diffie-Hellman key exchange.
+#[derive(Zeroize)]
+#[zeroize(drop)]
+pub struct SharedSecret(MontgomeryPoint);
+
+impl PublicKey {
+    /// Converts a bytes slice into a Public key
+    /// Returns None if:
+    /// -  The length of the slice is not 56
+    /// -  The point is a low order point
+    pub fn from_bytes(bytes: &[u8]) -> Option<PublicKey> {
+        let public_key = PublicKey::from_bytes_unchecked(bytes)?;
+        if public_key.0.is_low_order() {
+            return None;
+        }
+        Some(public_key)
+    }
+    /// Converts a bytes slice into a Public key
+    /// Returns None if:
+    /// -  The length of the slice is not 56
+    pub fn from_bytes_unchecked(bytes: &[u8]) -> Option<PublicKey> {
+        // First check if we have 56 bytes
+        if bytes.len() != 56 {
+            return None;
+        }
+
+        // Check if the point has low order
+        let arr = slice_to_array(bytes);
+        let point = MontgomeryPoint(arr);
+
+        Some(PublicKey(point))
+    }
+
+    /// Converts a public key into a byte slice
+    pub fn as_bytes(&self) -> &[u8; 56] {
+        self.0.as_bytes()
+    }
+}
+
+impl SharedSecret {
+    /// Converts a shared secret into a byte slice
+    pub fn as_bytes(&self) -> &[u8; 56] {
+        self.0.as_bytes()
+    }
+}
+
+impl Secret {
+    /// Generate a x448 `Secret` key.
+    // Taken from dalek-x25519
+    pub fn new<T>(csprng: &mut T) -> Self
+    where
+        T: RngCore + CryptoRng + ?Sized,
+    {
+        let mut bytes = [0u8; 56];
+
+        csprng.fill_bytes(&mut bytes);
+
+        Secret::from(bytes)
+    }
+
+    /// Clamps the secret key according to RFC7748
+    fn clamp(&mut self) {
+        self.0[0] &= 252;
+        self.0[55] |= 128;
+    }
+
+    /// Views a Secret as a Scalar
+    fn as_scalar(&self) -> MontgomeryScalar {
+        let secret = U448::from_le_slice(&self.0);
+        MontgomeryScalar::from_uint_unchecked(secret)
+    }
+
+    /// Performs a Diffie-hellman key exchange between the secret key and an external public key
+    pub fn as_diffie_hellman(&self, public_key: &PublicKey) -> Option<SharedSecret> {
+        // Check if the point is one of the low order points
+        if public_key.0.is_low_order() {
+            return None;
+        }
+        let shared_key = &public_key.0 * &self.as_scalar();
+        Some(SharedSecret(shared_key))
+    }
+
+    /// Performs a Diffie-hellman key exchange once between the secret key and an external public key
+    pub fn to_diffie_hellman(self, public_key: &PublicKey) -> Option<SharedSecret> {
+        self.as_diffie_hellman(public_key)
+    }
```

Comment:
This doesn't seem like a helpful distinction. I think we should just have the borrowing version. Even for ephemeral D-H keys, there are valid reasons to perform more than one D-H operation (i.e. 3DH)